src/events/guildAuditLogEntryCreate.ts

```diff
@@ -0,0 +1,224 @@
+import { AuditLogEvent, EmbedBuilder } from 'discord.js';
+import { sendEventLogMessage } from '@/util';
+import type { Guild, User, GuildAuditLogsEntry } from 'discord.js';
+
+export default async function guildAuditLogEntryCreateHandler(auditLogEntry: GuildAuditLogsEntry, guild: Guild): Promise<void> {
+	switch (auditLogEntry.action) { 
+		case AuditLogEvent.MemberUpdate: {
+
+			// * Unsure why just checking the action type doesn't narrow the type here, probably too much type inference going on?
+			const user = (auditLogEntry as unknown as GuildAuditLogsEntry<AuditLogEvent.MemberUpdate>).target;
+			if (!user) {
+				return;
+			}
+
+			const executor = auditLogEntry.executor;
+			if (!executor) {
+				return;
+			}
+
+			// * This checks that a user was timed out
+			const timeoutChange = auditLogEntry.changes.find(change => change.key === 'communication_disabled_until');
+			if (timeoutChange) {
+				// * The types here should always be correct, but I have no idea how to convince the discord.js types of that...
+				if (timeoutChange.new && timeoutChange.new && typeof timeoutChange.new === 'string') {
+					const timeout = new Date(timeoutChange.new);
+					await handleMemberTimedOut(guild, user, executor, auditLogEntry.reason, timeout);
+				}
+			}
+
+			// * This checks that a user has changed their name
+			const nickChange = auditLogEntry.changes.find(change => change.key === 'nick');
+			// * The types here should always be correct, but I have no idea how to convince the discord.js types of that...
+			if (nickChange && typeof nickChange.new === 'string' && (nickChange.old === undefined || typeof nickChange.old === 'string')) {
```

Comment:
Why the check for both `undefined` or `string`? That feels redundant, since it would always be one of those. It reads like checking `if (true || false)` to me

Also what isses with types were you having? TypeScript seems perfectly happy to narrow these types in my test:

![Screenshot from 2024-07-19 08-30-43](https://github.com/user-attachments/assets/41d3029d-2f8c-46b7-8ca0-63eee6f02a8d)

Reply:
`old` and `new` for a nickname change _should_ have the type `string | undefined` because of the nature of the change. But the type doesn't get narrowed even though we know what kind of change event it is, and since `AuditLogChange` isn't parameterised on a type, we can't specialise it any more.

That means without a cast or a guard, `old` and `new` have a type like this:
![image](https://github.com/user-attachments/assets/478e668c-cc5b-4306-bd45-1a786e2dca67)

You can even do things like:
```ts
		const x: AuditLogChange = {
			key: 'nick',
			old: 1,
			new: 2
		};
```
And that's _totally_ valid. Even though the types should be able to be narrowed. This is because `AuditLogChange` looks like this:

```ts
export interface AuditLogChange {
  key: APIAuditLogChange['key'];
  old?: APIAuditLogChange['old_value'];
  new?: APIAuditLogChange['new_value'];
}
```
Where `APIAuditLogChange` is the union type of all the different interfaces that an audit change can be.

I'm not even sure how we'd write a type guard for this because there's not even a value to compare, only types.

Reply:
Yeah I agree, Discord.JS's types are kinda fucked. To the point where I honestly sort of want to roll our own minimal client just for us with better types, because it's getting ridiculous. I've taken a look too now

Anyway, this check can still be cleaned up a bit by only checking the `typeof === 'string'`. If it's `undefined` then that will be false, which ensures that we have the correct values and that they exist:

```ts
import { type GuildAuditLogsEntry, AuditLogEvent } from 'discord.js';

function test(auditLogEntry: GuildAuditLogsEntry) {
	switch (auditLogEntry.action) {
		case AuditLogEvent.MemberUpdate:
			const nickChange = auditLogEntry.changes.find(change => change.key === 'nick');

			if (nickChange && typeof nickChange.old === 'string' && typeof nickChange.new === 'string') {
				test2(nickChange.old);
				test2(nickChange.new);
			}
			break;
	}
}

// Simulate a function that MUST have a string, not undefined/null
function test2(str: string) {

}
```

Reply:
In theory both of these should be able to be undefined because you can remove your server nick / add one without having one before, so in theory we need the undefined check for both

Reply:
Ah that's right, I forgot about that. In that case, this doesn't seem to allow for `new` to be undefined?

Reply:
It'd be pretty ugly and hard to read if I inlined this so I've added a type predicate to tidy things up:
```ts
if (nickChange && isStringOrUndefined(nickChange.old) && isStringOrUndefined(nickChange.new)) {
	await handleMemberNicknameChange(guild, user, nickChange.old, nickChange.new);
}

function isStringOrUndefined(input: any): input is string | undefined {
	return input === undefined || typeof input === 'string';
}
```

You ok with that?

Reply:
That's perfect :+1: